File: apo-core/trunk/src/main/java/com/uqbar/renascent/aop/pointcut/predicate/OrPredicate.java
package com.uqbar.renascent.aop.pointcut.predicate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javassist.CtClass;
import javassist.CtField;

/**
 * 
 */
public class OrPredicate implements APredicate {
	private List<APredicate> components = new ArrayList<APredicate>();

	public OrPredicate(APredicate... predicate) {
		this.components = Arrays.asList(predicate);
	}

	@Override
	public boolean evaluate(CtClass ctClass) {
		for (APredicate component : this.components) {
			if (component.evaluate(ctClass)) {
				return true;
			}
		}
		return false;
	}

	@Override
	public boolean evaluate(CtField ctField) {
		for (APredicate component : this.components) {
			if (component.evaluate(ctField)) {
				return true;
			}
		}
		return false;
	}

}

File: apo-core/trunk/src/main/scala/com/uqbar/aop/pointcut/predicate/OrPredicate.scala
package com.uqbar.aop.pointcut.predicate
import scala.collection.mutable.Buffer
import javassist.CtClass
import javassist.CtField

class OrPredicate extends APredicate {

  var components = Buffer[APredicate]();

  def this(predicate: APredicate*) {
    this()
    this.components = predicate.toBuffer
  }

  override def evaluate(ctClass: CtClass): Boolean = {
    this.components.foreach(component => {
      if (component.evaluate(ctClass)) {
        return true;
      }
    })
    return false;
  }

  override def evaluate(ctField: CtField): Boolean = {
    this.components.foreach(component => {
      if (component.evaluate(ctField)) {
        return true;
      }
    })
    return false;
  }

}